xtask/src/main.rs

```diff
@@ -5,148 +5,135 @@
-    /// given args.
-    fn new(matches: &clap::ArgMatches) -> BuildProfile {
-        if matches.contains_id("release") {
-            BuildProfile::Release
-        } else {
-            BuildProfile::Debug
-        }
-    }
-
-    /// Returns the subdirectory component corresponding
-    /// to the build type.
-    fn dir(self) -> &'static Path {
-        Path::new(match self {
-            Self::Debug => "debug",
-            Self::Release => "release",
-        })
+    // Returns the cargo argument corresponding to the given
+    // profile.
+    fn to_str(&self) -> &'static str {
+        self.release.then_some("--release").unwrap_or("")
     }
 
-    /// Yields the appropriate cargo argument for the given
-    /// build profile.
-    fn build_type(self) -> Option<&'static str> {
-        match self {
-            Self::Release => Some("--release"),
-            Self::Debug => None,
-        }
+    // Returns the output subdirectory component corresponding
+    // to the profile.
+    fn dir(&self) -> &'static Path {
+        Path::new(if self.release { "release" } else { "debug" })
     }
 }
 
-/// Build arguments including path to the compressed
-/// cpio archive we use as a "ramdisk
-#[derive(Clone, Debug)]
-struct BuildArgs {
-    profile: BuildProfile,
-    cpioz: String,
+/// Cargo `--locked` setting; separate from BuildProfile because
+/// `clippy` uses it but doesn't care about debug/release.
+#[derive(Parser)]
+struct Locked {
+    /// Build locked to Cargo.lock
+    #[clap(long)]
+    locked: bool,
 }
 
-impl BuildArgs {
-    /// Extracts the build profile type from the given matched
-    /// arguments.  Debug is the default.
-    fn new(matches: &clap::ArgMatches) -> BuildArgs {
-        let profile = BuildProfile::new(matches);
-        let cpioz = matches.get_one::<String>("cpioz").unwrap().to_string();
-        BuildArgs { profile, cpioz }
+impl Locked {
+    fn to_str(&self) -> &str {
+        self.locked.then_some("--locked").unwrap_or("")
     }
 }
 
 fn main() {
-    let matches = parse_args();
-    match matches.subcommand() {
-        Some(("build", m)) => build(BuildArgs::new(m), m.contains_id("locked")),
-        Some(("test", m)) => {
-            test(BuildProfile::new(m), m.contains_id("locked"))
+    let xtask = Xtask::parse();
+    match xtask.cmd {
+        Command::Build { profile, locked, cpioz } => {
+            build(profile, locked, cpioz)
         }
-        Some(("disasm", m)) => disasm(
-            BuildArgs::new(m),
-            m.contains_id("locked"),
-            m.contains_id("source").then_some("-S"),
-        ),
-        Some(("expand", _m)) => expand(),
-        Some(("clippy", m)) => clippy(m.contains_id("locked")),
-        Some(("clean", _m)) => clean(),
-        _ => {
-            println!("Unknown command");
-            process::exit(1);
+        Command::Test { profile, locked } => test(profile, locked),
+        Command::Disasm { profile, locked, source, cpioz } => {
```

Comment:
FWIW, if this becomes annoying (unpacking the bits and then passing them), you could also make `Disasm` contain only a `DisasmArgs` struct that you pass into the `disasm` function.

For four things, it's probably not worth it, but it can help keep track of things if you sprout more.

Reply:
Thanks; that was something I was going to ask about. For this, I agree it's not necessary, but it sure struck me that it could get pretty nasty pretty fast and I wasn't sure if the struct-in-enum pattern was the best, but it seems like it. I presume one would flatten that at the clap level?